### ISSUE TITLE
Generator return type is not detected if there's no yields outside of arguments lists

### ISSUE THREAD

**Issue by remorhaz** [vimeo/psalm#4122]
I've discovered this problem while working with promises of [Amp framework](https://amphp.org/), but I've made a simplified example that reproduces it.

The `Foo::a()` method creates a generator that returns `int`, but Psalm detects that it returns `mixed`:
```php
/**
 * @psalm-return Generator<mixed, mixed, mixed, int>
 */
public function a(): Generator
{
    return $this->b(yield $this->c());
}
```
https://psalm.dev/r/3ce39a48c3

The problem doesn't occur if we put at least one yield instruction directly into generator's body, like this:
```php
public function a(): Generator
{
    yield $this->c();
    return $this->b(yield $this->c());
}
```
